File: src/c/gpio.c
#include <stdlib.h>
#include <stdio.h>

#include "gpio.h"

#define GPIO_MODE_FILE_FORMAT "/sys/devices/virtual/misc/gpio/mode/gpio%d"
#define GPIO_PIN_FILE_FORMAT "/sys/devices/virtual/misc/gpio/pin/gpio%d"

static char name_buff[128];

void set_gpio_mode(int pin,int mode)
{
    FILE* f_mode;
    
    sprintf(name_buff,GPIO_MODE_FILE_FORMAT,pin);
    f_mode=fopen(name_buff,"w");
    if(f_mode==NULL)
    {
        fprintf(stderr,"Unable to open gpio mode file: %s\n",name_buff);
        return;
    }
    
    fwrite((mode==LOW)? "0":"1",1,1,f_mode);
    
    fflush(f_mode);
    fclose(f_mode);
}

int get_gpio_pin(int pin)
{
    FILE* f_pin;
    int c;
    
    sprintf(name_buff,GPIO_PIN_FILE_FORMAT,pin);
    f_pin=fopen(name_buff,"r");
    if(f_pin==NULL)
    {
        fprintf(stderr,"Unable to open gpio pin file: %s\n",name_buff);
        return -1;
    }
    
    c=fgetc(f_pin);
    fclose(f_pin);
    
    if(c=='0')
    {
        return LOW;
    }
    else if(c=='1')
    {
        return HIGH;
    }
    else
    {
        return -1;
    }
}

void set_gpio_pin(int pin,int level)
{
    FILE* f_pin;
    
    sprintf(name_buff,GPIO_PIN_FILE_FORMAT,pin);
    f_pin=fopen(name_buff,"w");
    if(f_pin==NULL)
    {
        fprintf(stderr,"Unable to open gpio pin file: %s\n",name_buff);
        return;
    }
    
    fwrite((level==LOW)? "0":"1",1,1,f_pin);
    
    fflush(f_pin);
    fclose(f_pin);
}

File: src/python/gpio.py
import sys

HIGH=1
LOW =0

OUTPUT=1
INPUT =0

GPIO_MODE_FILE_FORMAT="/sys/devices/virtual/misc/gpio/mode/gpio%d"
GPIO_PIN_FILE_FORMAT="/sys/devices/virtual/misc/gpio/pin/gpio%d"

def set_gpio_mode(pin,mode):
    name=GPIO_MODE_FILE_FORMAT%pin
    ret=True
    try:
        f_mode=open(name,"w")
	    f_mode.write(str(mode))
    	f_mode.flush()
    except IOError as e:
        sys.stderr.write(str(e)+"\n")
        ret=False
    
    f_mode.close()
    return ret

def get_gpio_pin(pin):
    name=GPIO_PIN_FILE_FORMAT%pin
    try:
        f_pin=open(name,"r");
    except IOError as e:
        sys.stderr.write(str(e)+"\n")
        return -1
    
    c=f_pin.read(1)
    
    f_pin.close()
    
    if(c=="0"):
        return LOW;
    elif(c=="1"):
        return HIGH;
    else:
        return -1;

def set_gpio_pin(pin,level):
    name=GPIO_PIN_FILE_FORMAT%pin
    ret=True
    try:
        f_pin=open(name,"w")
        f_pin.write(str(level))
        f_pin.flush();
    except IOError as e:
        sys.stderr.write(str(e)+"\n")
        ret=False
    
    f_pin.close();
    return ret
